File: Top Interview 150/Java Code/Binary Search Tree/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    int cnt=0;
    void inorder(TreeNode root,int k){
        if(root==null) return;
        inorder(root.left,k);
        cnt++;
        if(cnt==k){
            ans=root.val;
            return;
        }
        inorder(root.right,k);
    }
    
    public int kthSmallest(TreeNode root, int k) {
         inorder(root,k);
        return ans;   
    }
}

 /**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// Iterative methos using stack.
class Solution {
    
    public int kthSmallest(TreeNode root, int k) {
         Stack<TreeNode>st=new Stack<>();
        while(true){
            while(root!=null){
                st.add(root);
                root=root.left;
            }
            root=st.pop();
             if(--k==0) return root.val;
             root=root.right;
        }
        
    }
}

 

File: Top Interview 150/C++ Code/Binary Search Tree/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// using a vector.
class Solution {
public:
   
    void kth(TreeNode *root,vector<int>&v){
        if(root->left!=NULL) kth(root->left,v);
        v.push_back(root->val);
        if(root->right!=NULL) kth(root->right,v);
    }
    int kthSmallest(TreeNode* root, int k) {
        vector<int>v;
       kth(root,v);
        return v[k-1];
    }
};


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// using without any vector
class Solution2 {
public:
    int ans;
    void inorder(TreeNode *root,int &k){
        if(!root) return;
        inorder(root->left,k);
        k--;
        if(k==0){
            ans=root->val;
            return;
        }
        inorder(root->right,k);
    }
        
          
    int kthSmallest(TreeNode* root, int k) {
         inorder(root,k);
        return ans;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//using iteration method

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode *>st;
        while(root!=NULL || !st.empty()){
            while(root!=NULL){
                st.push(root);
                root=root->left;
            }
            root=st.top();
            st.pop();
            if(--k==0) return root->val;
            root=root->right;
        }
        return 0;
    }
};
